File: core/geometry.py
"""Геометрические операции: упрощение, офсет, сортировка, масштабирование, мостики."""

import math

try:
    from shapely.geometry import LineString
    HAS_SHAPELY = True
except ImportError:
    HAS_SHAPELY = False


# ─────────────────────────── Упрощение Рамера-Дугласа-Пекера ────────────────

def _rdp(points, epsilon):
    if len(points) < 3:
        return points
    start, end = points[0], points[-1]
    dx, dy = end[0] - start[0], end[1] - start[1]
    dist_sq = dx * dx + dy * dy
    max_dist, idx = 0.0, 0
    for i in range(1, len(points) - 1):
        px, py = points[i]
        if dist_sq == 0:
            d = math.hypot(px - start[0], py - start[1])
        else:
            t = ((px - start[0]) * dx + (py - start[1]) * dy) / dist_sq
            t = max(0.0, min(1.0, t))
            d = math.hypot(px - (start[0] + t * dx), py - (start[1] + t * dy))
        if d > max_dist:
            max_dist, idx = d, i
    if max_dist > epsilon:
        left = _rdp(points[:idx + 1], epsilon)
        right = _rdp(points[idx:], epsilon)
        return left[:-1] + right
    return [start, end]


def simplify_chain(chain, epsilon):
    """Упрощает контур алгоритмом Рамера-Дугласа-Пекера.

    Args:
        chain: Список точек [(x,y), ...]
        epsilon: Допуск упрощения

    Returns:
        list: Упрощенный контур
    """
    if epsilon < 0:
        raise ValueError(f"Допуск упрощения должен быть >= 0, получено: {epsilon}")
    if epsilon <= 0 or len(chain) < 3:
        return chain
    return _rdp(chain, epsilon)


# ─────────────────────────── Chaikin-сглаживание ────────────────────────────

def chaikin_smooth(chain, passes, closed=True):
    """Сглаживает контур алгоритмом Chaikin (итеративное срезание углов).

    Каждый проход: Q = 0.75·P[i] + 0.25·P[i+1],  R = 0.25·P[i] + 0.75·P[i+1].
    Сходится к B-сплайну 2-го порядка.

    Рекомендации: 1–2 прохода — лёгкое сглаживание, 3–4 — оптимально,
    5–6 — очень гладко (мелкие детали округляются).

    Args:
        chain: Список точек [(x,y), ...]. Может быть замкнутым (первая==последняя).
        passes: Число проходов (0 = нет сглаживания).
        closed: True если контур замкнутый.

    Returns:
        list: Сглаженный контур. Если closed=True, последняя точка == первой.

    Raises:
        ValueError: Если passes < 0.
    """
    if passes < 0:
        raise ValueError(f"Число проходов должно быть >= 0, получено: {passes}")
    if passes == 0 or len(chain) < 3:
        return list(chain)
    p = list(chain)
    if closed and len(p) > 1:
        x0, y0 = p[0]
        xl, yl = p[-1]
        if math.hypot(x0 - xl, y0 - yl) < 1e-9:
            p = p[:-1]
    for _ in range(passes):
        n = len(p)
        nxt = []
        for j in range(n):
            ax, ay = p[j]
            bx, by = p[(j + 1) % n]
            nxt.append((0.75 * ax + 0.25 * bx, 0.75 * ay + 0.25 * by))
            nxt.append((0.25 * ax + 0.75 * bx, 0.25 * ay + 0.75 * by))
        p = nxt
    if closed:
        p.append(p[0])
    return p


# ─────────────────────────── Ре-сэмплинг по длине дуги ─────────────────────

def resample_by_length(chain, step_len, closed=True):
    """Расставляет точки вдоль контура с равным шагом step_len.

    Равномерный ре-сэмплинг: плотность G1-сегментов одинакова везде —
    на прямых и на кривых. В отличие от RDP, который оставляет много точек
    только в местах изломов.

    Практический выбор: 0.5–1.5 пикс. для хорошего разрешения.

    Args:
        chain: Список точек [(x,y), ...]. Может быть замкнутым.
        step_len: Расстояние между соседними точками (те же единицы, что и цепочка).
        closed: True если контур замкнутый.

    Returns:
        list: Ре-сэмплированный контур.

    Raises:
        ValueError: Если step_len <= 0.
    """
    if step_len <= 0:
        raise ValueError(f"Шаг ре-сэмплинга должен быть > 0, получено: {step_len}")
    if len(chain) < 2:
        return list(chain)
    src = list(chain)
    if closed and len(src) > 1:
        x0, y0 = src[0]
        xl, yl = src[-1]
        if math.hypot(x0 - xl, y0 - yl) < 1e-9:
            src = src[:-1]
    if closed:
        src = src + [src[0]]
    result = [src[0]]
    acc = 0.0
    for i in range(1, len(src)):
        dx = src[i][0] - src[i - 1][0]
        dy = src[i][1] - src[i - 1][1]
        seg_len = math.hypot(dx, dy)
        if seg_len < 1e-9:
            continue
        traveled = 0.0
        while acc + (seg_len - traveled) >= step_len:
            move = step_len - acc
            traveled += move
            t = traveled / seg_len
            result.append((src[i - 1][0] + t * dx, src[i - 1][1] + t * dy))
            acc = 0.0
        acc += seg_len - traveled
    if closed:
        result.append(result[0])
    return result


# ─────────────────────────── Офсет (tool compensation) ──────────────────────

def offset_chain(chain, offset):
    """Смещает контур на указанное расстояние (компенсация инструмента).
    
    Args:
        chain: Список точек [(x,y), ...]
        offset: Смещение (отрицательное = внутрь, положительное = наружу)
    
    Returns:
        list: Список смещенных контуров
    """
    if not HAS_SHAPELY or offset == 0:
        return [chain]
    try:
        ls = LineString(chain)
        result = ls.offset_curve(offset)
        if result is None or result.is_empty:
            return [chain]
        if result.geom_type == "LineString":
            return [list(result.coords)]
        elif result.geom_type == "MultiLineString":
            return [list(g.coords) for g in result.geoms]
    except Exception:
        pass
    return [chain]


# ─────────────────────────── Масштабирование ────────────────────────────────

def get_bounds(chains):
    all_pts = [pt for c in chains for pt in c]
    if not all_pts:
        return (0, 0, 0, 0)
    xs = [p[0] for p in all_pts]
    ys = [p[1] for p in all_pts]
    return (min(xs), min(ys), max(xs), max(ys))


def scale_chains(chains, target_w=None, target_h=None, keep_aspect=True):
    """Масштабирует контуры к целевым размерам.
    
    Args:
        chains: Список контуров
        target_w: Целевая ширина (мм)
        target_h: Целевая высота (мм)
        keep_aspect: Сохранять ли пропорции
    
    Returns:
        list: Масштабированные контуры
    """
    if not chains or (target_w is None and target_h is None):
        return chains
    
    # Валидация размеров
    if target_w is not None and target_w <= 0:
        raise ValueError(f"Целевая ширина должна быть > 0, получено: {target_w}")
    if target_h is not None and target_h <= 0:
        raise ValueError(f"Целевая высота должна быть > 0, получено: {target_h}")
    
    min_x, min_y, max_x, max_y = get_bounds(chains)
    src_w = max_x - min_x
    src_h = max_y - min_y
    if src_w == 0 or src_h == 0:
        return chains
    if target_w is not None and target_h is not None:
        sx = target_w / src_w
        sy = target_h / src_h
        if keep_aspect:
            sx = sy = min(sx, sy)
    elif target_w is not None:
        sx = sy = target_w / src_w
    else:
        sx = sy = target_h / src_h
    result = []
    for chain in chains:
        scaled = [((x - min_x) * sx, (y - min_y) * sy) for x, y in chain]
        result.append(scaled)
    return result


# ─────────────────────────── Сортировка контуров (Nearest Neighbor) ─────────

def sort_chains_nearest(chains):
    if len(chains) <= 1:
        return chains
    remaining = list(range(len(chains)))
    ordered = [remaining.pop(0)]
    while remaining:
        last = chains[ordered[-1]][-1]
        best_idx = 0
        best_dist = float("inf")
        for i, ri in enumerate(remaining):
            sx, sy = chains[ri][0]
            d = (sx - last[0]) ** 2 + (sy - last[1]) ** 2
            if d < best_dist:
                best_dist = d
                best_idx = i
        ordered.append(remaining.pop(best_idx))
    return [chains[i] for i in ordered]


# ─────────────────────────── Мостики (tabs) ─────────────────────────────────

def _chain_length(chain):
    return sum(math.hypot(chain[i + 1][0] - chain[i][0], chain[i + 1][1] - chain[i][1])
               for i in range(len(chain) - 1))


def insert_bridges(chain, bridge_size, num_bridges=2):
    """Вставляет мостики (perемычки) в контур.
    
    Args:
        chain: Список точек [(x,y), ...]
        bridge_size: Размер мостика (мм)
        num_bridges: Количество мостиков
    
    Returns:
        list: Список сегментов [(segment, is_bridge), ...]
    
    Raises:
        ValueError: При некорректных параметрах
    """
    if bridge_size <= 0:
        raise ValueError(f"Размер мостика должен быть > 0, получено: {bridge_size}")
    if num_bridges < 1:
        raise ValueError(f"Количество мостиков должно быть >= 1, получено: {num_bridges}")
    if len(chain) < 2:
        raise ValueError("Контур должен иметь минимум 2 точки")
    
    total = _chain_length(chain)
    if total < bridge_size * num_bridges * 2 or bridge_size <= 0:
        return [(chain, False)]

    interval = total / num_bridges
    bridge_starts = [interval * i + interval * 0.5 for i in range(num_bridges)]

    segments = []
    cur_dist = 0.0
    seg_pts = [chain[0]]
    bridge_idx = 0
    in_bridge = False

    for i in range(len(chain) - 1):
        p0, p1 = chain[i], chain[i + 1]
        step = math.hypot(p1[0] - p0[0], p1[1] - p0[1])
        if step == 0:
            continue

        events = []
        if bridge_idx < len(bridge_starts):
            bs = bridge_starts[bridge_idx]
            be = bs + bridge_size
            for dist_event in [bs, be]:
                d_local = dist_event - cur_dist
                if 0 < d_local < step:
                    t = d_local / step
                    events.append((t, dist_event))

        events.sort()
        for (t, dist_event) in events:
            interp = (p0[0] + t * (p1[0] - p0[0]), p0[1] + t * (p1[1] - p0[1]))
            seg_pts.append(interp)
            segments.append((seg_pts, in_bridge))
            seg_pts = [interp]
            in_bridge = not in_bridge
            if not in_bridge:
                bridge_idx += 1

        seg_pts.append(p1)
        cur_dist += step

    if len(seg_pts) >= 2:
        segments.append((seg_pts, in_bridge))

    return segments

File: Html/js/geometry.js
/**
 * Geometry — Геометрические операции
 * Аналог core/geometry.py
 * 
 * Реализует:
 * - Упрощение контуров RDP
 * - Офсет контуров (tool compensation)
 * - Масштабирование контуров
 * - Сортировка контуров (nearest neighbor)
 * - Вставку мостиков (bridges)
 */

class Geometry {
    /**
     * Упрощает контур алгоритмом Рамера-Дугласа-Пекера
     * @param {Array<[number, number]>} chain 
     * @param {number} epsilon 
     * @returns {Array<[number, number]>}
     */
    static simplifyChain(chain, epsilon) {
        if (epsilon <= 0 || chain.length < 3) return chain;
        return Geometry._rdp(chain, epsilon);
    }

    /**
     * RDP алгоритм (рекурсивный)
     */
    static _rdp(points, epsilon) {
        if (points.length < 3) return points;
        
        const start = points[0];
        const end = points[points.length - 1];
        const dx = end[0] - start[0];
        const dy = end[1] - start[1];
        const distSq = dx * dx + dy * dy;
        
        let maxDist = 0;
        let idx = 0;
        
        for (let i = 1; i < points.length - 1; i++) {
            const px = points[i][0];
            const py = points[i][1];
            
            let d;
            if (distSq === 0) {
                d = Math.hypot(px - start[0], py - start[1]);
            } else {
                let t = ((px - start[0]) * dx + (py - start[1]) * dy) / distSq;
                t = Math.max(0, Math.min(1, t));
                d = Math.hypot(px - (start[0] + t * dx), py - (start[1] + t * dy));
            }
            
            if (d > maxDist) {
                maxDist = d;
                idx = i;
            }
        }
        
        if (maxDist > epsilon) {
            const left = Geometry._rdp(points.slice(0, idx + 1), epsilon);
            const right = Geometry._rdp(points.slice(idx), epsilon);
            return [...left.slice(0, -1), ...right];
        }
        
        return [start, end];
    }

    /**
     * Смещает контур на указанное расстояние (компенсация инструмента)
     * Упрощённая реализация без Shapely — просто масштабируем контур от центра
     * @param {Array<[number, number]>} chain 
     * @param {number} offset 
     * @returns {Array<Array<[number, number]>>}
     */
    static offsetChain(chain, offset) {
        if (offset === 0 || chain.length < 3) return [chain];
        
        // Упрощённый офсет: находим центр и масштабируем
        // Это НЕ точный офсет как у Shapely, но работает для простых контуров
        const center = Geometry._centroid(chain);
        
        const result = chain.map(([x, y]) => {
            const dx = x - center[0];
            const dy = y - center[1];
            const dist = Math.hypot(dx, dy);
            if (dist === 0) return [x, y];
            
            const scale = (dist + offset) / dist;
            return [center[0] + dx * scale, center[1] + dy * scale];
        });
        
        return [result];
    }

    /**
     * Вычисление центроида полигона
     */
    static _centroid(points) {
        let cx = 0, cy = 0;
        const n = points.length;
        for (const [x, y] of points) {
            cx += x;
            cy += y;
        }
        return [cx / n, cy / n];
    }

    /**
     * Получить bounding box контуров
     * @param {Array<Array<[number, number]>>} chains 
     * @returns {[number, number, number, number]}
     */
    static getBounds(chains) {
        const allPts = chains.flat();
        if (allPts.length === 0) return [0, 0, 0, 0];
        
        let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
        for (const [x, y] of allPts) {
            if (x < minX) minX = x;
            if (y < minY) minY = y;
            if (x > maxX) maxX = x;
            if (y > maxY) maxY = y;
        }
        
        return [minX, minY, maxX, maxY];
    }

    /**
     * Масштабирует контуры к целевым размерам
     * @param {Array<Array<[number, number]>>} chains 
     * @param {number|null} targetW 
     * @param {number|null} targetH 
     * @param {boolean} keepAspect 
     * @returns {Array<Array<[number, number]>>}
     */
    static scaleChains(chains, targetW = null, targetH = null, keepAspect = true) {
        if (!chains || chains.length === 0 || (targetW === null && targetH === null)) {
            return chains;
        }
        
        const [minX, minY, maxX, maxY] = Geometry.getBounds(chains);
        const srcW = maxX - minX;
        const srcH = maxY - minY;
        
        if (srcW === 0 || srcH === 0) return chains;
        
        let sx, sy;
        if (targetW !== null && targetH !== null) {
            sx = targetW / srcW;
            sy = targetH / srcH;
            if (keepAspect) sx = sy = Math.min(sx, sy);
        } else if (targetW !== null) {
            sx = sy = targetW / srcW;
        } else {
            sx = sy = targetH / srcH;
        }
        
        return chains.map(chain => 
            chain.map(([x, y]) => [(x - minX) * sx, (y - minY) * sy])
        );
    }

    /**
     * Сортировка контуров методом ближайшего соседа
     * @param {Array<Array<[number, number]>>} chains 
     * @returns {Array<Array<[number, number]>>}
     */
    static sortChainsNearest(chains) {
        if (chains.length <= 1) return chains;
        
        const remaining = Array.from({ length: chains.length }, (_, i) => i);
        const ordered = [remaining.shift()];
        
        while (remaining.length > 0) {
            const lastChain = chains[ordered[ordered.length - 1]];
            const last = lastChain[lastChain.length - 1];
            
            let bestIdx = 0;
            let bestDist = Infinity;
            
            for (let i = 0; i < remaining.length; i++) {
                const ri = remaining[i];
                const first = chains[ri][0];
                const d = (first[0] - last[0]) ** 2 + (first[1] - last[1]) ** 2;
                if (d < bestDist) {
                    bestDist = d;
                    bestIdx = i;
                }
            }
            
            ordered.push(remaining.splice(bestIdx, 1)[0]);
        }
        
        return ordered.map(i => chains[i]);
    }

    /**
     * Вычисляет длину полилинии
     * @param {Array<[number, number]>} chain 
     * @returns {number}
     */
    static chainLength(chain) {
        let total = 0;
        for (let i = 0; i < chain.length - 1; i++) {
            total += Math.hypot(chain[i + 1][0] - chain[i][0], chain[i + 1][1] - chain[i][1]);
        }
        return total;
    }

    /**
     * Вставляет мостики (perемычки) в контур
     * Аналог insert_bridges в geometry.py
     * @param {Array<[number, number]>} chain 
     * @param {number} bridgeSize 
     * @param {number} numBridges 
     * @returns {Array<{segment: Array<[number, number]>, isBridge: boolean}>}
     */
    static insertBridges(chain, bridgeSize, numBridges = 2) {
        if (chain.length < 2) return [{ segment: chain, isBridge: false }];
        
        const total = Geometry.chainLength(chain);
        if (total < bridgeSize * numBridges * 2 || bridgeSize <= 0) {
            return [{ segment: chain, isBridge: false }];
        }
        
        const interval = total / numBridges;
        const bridgeStarts = Array.from({ length: numBridges }, (_, i) => 
            interval * i + interval * 0.5
        );
        
        const segments = [];
        let curDist = 0;
        let segPts = [chain[0]];
        let bridgeIdx = 0;
        let inBridge = false;
        
        for (let i = 0; i < chain.length - 1; i++) {
            const p0 = chain[i];
            const p1 = chain[i + 1];
            const step = Math.hypot(p1[0] - p0[0], p1[1] - p0[1]);
            if (step === 0) continue;
            
            const events = [];
            if (bridgeIdx < bridgeStarts.length) {
                const bs = bridgeStarts[bridgeIdx];
                const be = bs + bridgeSize;
                
                for (const distEvent of [bs, be]) {
                    const dLocal = distEvent - curDist;
                    if (dLocal > 0 && dLocal < step) {
                        const t = dLocal / step;
                        events.push({ t, dist: distEvent });
                    }
                }
            }
            
            events.sort((a, b) => a.t - b.t);
            
            for (const event of events) {
                const interp = [
                    p0[0] + event.t * (p1[0] - p0[0]),
                    p0[1] + event.t * (p1[1] - p0[1])
                ];
                segPts.push(interp);
                segments.push({ segment: [...segPts], isBridge });
                segPts = [interp];
                inBridge = !inBridge;
                if (!inBridge) bridgeIdx++;
            }
            
            segPts.push(p1);
            curDist += step;
        }
        
        if (segPts.length >= 2) {
            segments.push({ segment: segPts, isBridge });
        }
        
        return segments;
    }
}
